code/hdb/startup.q

```diff
@@ -1,4 +1,55 @@
 
+//called after new HDB cluster finishes loading all dependencies
+//triggers RDB reload instead of the HDB and removes old HDB cluster if it is a replacement
+.hdb.startup:{
+  gws:exec w from .servers.SERVERS where proctype = `gateway, .dotz.liveh w;
+  if[not count gws; :()]; // TODO : Handle case when none
+
+  // block queries to new hdb
+  gws @\: (`.gw.setserveractiveflag;.proc.procname;0b);
+  // wait for status change
+  .finspace.checkstatus[(.finspace.getcluster;.proc.procname);("RUNNING";"FAILURE");00:01;0D01];
+
+  // identify the row and retry connection
+  {[h;tgt]
+     tgtidx:first @[h;({exec i from .servers.SERVERS where procname=x};tgt);()];
+     if[null tgtidx; :()];
+     doretry:@[h;({null .servers.SERVERS[x][`w]};tgtidx);1b];
+     .lg.o[`hdbstartup;"will force retry row ",(-3!tgtidx)," in servers.SERVERS for gw handle ",(-3!h)," : ",-3!doretry];
+     if[doretry; h(`.servers.retryrows;tgtidx)];
+   }[;.proc.procname] each gws;
+
+  rdbs:select procname,w from .servers.SERVERS where proctype = `rdb, .dotz.liveh w;
+  // block queries to the rdbs
+  rdbnames:exec procname from rdbs;
+  gws @\: (`.gw.setserveractiveflag;rdbnames;0b);
```

Comment:
we're doing this before making the new hdb available - so doesn't this mean that a chunk of data will suddenly be unavailable (even though it was available via rdb prior to this point)? i.e. a query that runs now against hdb & rdb will return less data than it did a few seconds ago

I think this bit needs to happen simultaneously with the deregistering of old hbd and registering of new hdb. So sequence would be something like

1. new hdb starts and does it's initialisation etc.
2. once ready, new hdb signals gateway
3. gateway deregisters both old hdb & rdb, and registers new hdb (all at same time)
4. new hdb signals rdb to do "reload" (i.e. clearing old data)
5. rdb clears data & signals gateway that it is active again

I think this still means some data will be missing - any data received by rdb after the changeset for the new hdb was created will obviously not be in new hdb and won't be available from rdb while deregistered, but this is a smaller amount of data than everything from rdb. I can't think of a way around this without moving to having new rdb(s) as well as hdb(s).